### PR TITLE
`chunk_by_view`'s helper lambda returns `bool`

### LINKED ISSUE #2889: `<ranges>`: `chunk_by_view`'s helper lambda does not specify return type
https://github.com/microsoft/STL/blob/ef62d3fa0b8e4e2406b9bb74e916e1ca8a1df802/stl/inc/ranges#L6258-L6260
https://github.com/microsoft/STL/blob/ef62d3fa0b8e4e2406b9bb74e916e1ca8a1df802/stl/inc/ranges#L6272-L6274

This is basically the same problem as #2888.
I don't know how to use MSVC-STL-trunk with Compiler Explorer, but the following code should be rejected incorrectly.

```cpp
#include <ranges>

struct Bool {
  Bool& operator!();
  Bool() = default;
  Bool(const Bool&) = delete;
  operator bool() { return true; };
};

int main() {
  Bool b;
  int x[] = {42};
  auto r = std::views::iota(0, 5) | std::views::chunk_by([&](auto, auto) -> Bool& { return b; });
  r.begin();
}
```

### PR DESCRIPTION
**Pull request by cpplearner**
Fixes #2889

Casey's drive-by: Give up on the Standard's dream that we can avoid defining `projected::operator*`. There are just too many interactions with the "need for constant evaluation" rules and the use of types with internal linkage - it's not worth fighting anymore.

### CODE REVIEW

File: stl/inc/ranges [hunk 1]
```diff
@@ -6255,7 +6255,7 @@ namespace ranges {
             _STL_VERIFY(_Pred, "cannot increment a chunk_by_view iterator whose parent view has no predicate");
 #endif // _ITERATOR_DEBUG_LEVEL != 0
 
-            const auto _Not_pred = [&_Orig_pred = *_Pred]<class _Ty1, class _Ty2>(_Ty1&& _Left, _Ty2&& _Right) {
+            const auto _Not_pred = [&_Orig_pred = *_Pred]<class _Ty1, class _Ty2>(_Ty1&& _Left, _Ty2&& _Right) -> bool {
```
Comment (miscco) [inline stl/inc/ranges:6258]: I am wondering whether it would be worth it to use a simple struct with a templated call operator to avoid the potential bloat from the various lambda invocations

**Review by miscco — COMMENTED** [review]
We are missing some additional test coverage.

File: stl/inc/ranges (hunk 1 above)
Comment (strega-nil-ms) [reply stl/inc/ranges:6258]: I do not understand why we aren't using `not_fn` here...

Comment (cpplearner) [reply stl/inc/ranges:6258]: > I do not understand why we aren't using `not_fn` as the standard suggests here...

I don't want to drag in `<functional>`, and the use of `not_fn(bind(ref(*pred_), _2, _1))` (in the reference implementation of `find-prev` in [p2443](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2021/p2443r1.html#pnum_16)) makes me uneasy. I believe that by using one lambda, instead of three call wrappers, we can avoid a lot of forwarding.

**Review by strega-nil-ms — DISMISSED** [review]

**Review by strega-nil-ms — COMMENTED** [review]
Removing my approval as I do not believe it correctly implements the standard.

File: stl/inc/ranges [hunk 2]
```diff
@@ -6171,6 +6171,26 @@ namespace ranges {
         inline constexpr _Slide_fn slide;
     } // namespace views
 
+    template <class _Pr>
+    struct _Negated_pred {
+        template <class _Ty1, class _Ty2>
+        _NODISCARD constexpr bool operator()(_Ty1&& _Left, _Ty2&& _Right) {
```
Comment (strega-nil-ms) [inline stl/inc/ranges:6177]: Given https://github.com/microsoft/STL/issues/2889#issuecomment-1191736156, this should instead be

```suggestion
        _NODISCARD constexpr decltype(auto) operator()(_Ty1&& _Left, _Ty2&& _Right) {
```

File: stl/inc/ranges [hunk 3]
```diff
@@ -6171,6 +6171,26 @@ namespace ranges {
...
+            return !_STD invoke(_Pred, _STD forward<_Ty1>(_Left), _STD forward<_Ty2>(_Right));
+        }
+
+        _Pr& _Pred;
+    };
+
+    template <class _Pr>
+    struct _Backward_negated_pred {
+        template <class _Ty1, class _Ty2>
+        _NODISCARD constexpr bool operator()(_Ty1&& _Left, _Ty2&& _Right) {
```
Comment (strega-nil-ms) [inline stl/inc/ranges:6187]:
```suggestion
        _NODISCARD constexpr decltype(auto) operator()(_Ty1&& _Left, _Ty2&& _Right) {
```

**Review by strega-nil-ms — CHANGES_REQUESTED** [review]

File: tests/std/tests/P2443R1_views_chunk_by/test.cpp [hunk 4]
```diff
@@ -231,6 +231,24 @@ constexpr void instantiation_test() {
...
+struct Bool {
+    Bool() {}
+    Bool(const Bool&) = delete;
+    Bool& operator!() {
+        return *this;
+    }
+    operator bool() {
+        return true;
+    }
+};
```
Comment (JMazurkiewicz) [inline tests/std/tests/P2443R1_views_chunk_by/test.cpp:243]: This type does not model `boolean-testable`, so it cannot be used as the result of binary predicate.

[`[concept.booleantestable]/7`](http://eel.is/c++draft/concept.booleantestable) says:
> Let `e` be an expression such that `decltype((e))` is `T`. `T` models *`boolean-testable`* only if `bool(e) == !bool(!e)`.

[https://godbolt.org/z/oGeefGW6G](https://godbolt.org/z/oGeefGW6G)

File: stl/inc/ranges (hunk 2 above)
Comment (cpplearner) [reply stl/inc/ranges:6177]: As @hewillk mentions in #2889, the result is converted to `bool` in `adjacent_find` anyway, so returning `bool` should have the same end result as returning `decltype(auto)`.

An argument against `decltype(auto)` is: if the return type is `decltype(auto)`, the function body will need to be instantiated for `invoke_result_t` to get the actual return type, which is extra work.

(I also considered `decltype(!_STD invoke(_Pred, _STD forward<_Ty1>(_Left), _STD forward<_Ty2>(_Right)))`, but that's ugly and unnecessary.)

File: tests/std/tests/P2443R1_views_chunk_by/test.cpp (hunk 4 above)
Comment (cpplearner) [reply tests/std/tests/P2443R1_views_chunk_by/test.cpp:243]: I hope I can get away with it because the function is not executed.

If conformance to the semantic requirements is desired, I can make `operator!` return `std::cout`. 😈

File: stl/inc/ranges (hunk 2 above)
Comment (hewillk) [reply stl/inc/ranges:6177]: Returning `delctype(auto)` is verbose and unnecessary. Returning `bool` better reflects the characteristics of the "predicate".

Comment (strega-nil-ms) [reply stl/inc/ranges:6177]: mmh, okay. Then this needs to `static_cast<bool>(...)` in order to avoid explicit and implicit conversions resulting in different values;

```cxx
template <class Pr>
constexpr bool if_(Pr pr) {
    if (pr()) {
        return true;
    } else {
        return false;
    }
}

template <class Pr>
constexpr bool impl(Pr pr) {
    return pr();
}

template <class Pr>
constexpr bool check_if_impl(Pr pr) {
    return if_(pr) == impl(pr);
}

struct Bool {
    constexpr explicit operator bool() {
        return true;
    }
    constexpr operator int() {
        return 0;
    }
};

int main() {
    static_assert(check_if_impl([] { return Bool{}; })); // results in a failure
}
```

Comment (hewillk) [reply stl/inc/ranges:6177]: A `Bool` with only `explicit operator bool` does not model _`boolean-testable`_, which means we shouldn't do an explicit conversion here.

Comment (strega-nil-ms) [reply stl/inc/ranges:6177]: This _is_ implicitly convertible to `bool`, it just goes through a conversion to `int` first.

**Review by strega-nil-ms — APPROVED** [review]
It definitely took a bit to convince me that this is correct, but I have now been convinced.

File: tests/std/tests/P2443R1_views_chunk_by/test.cpp (hunk 4 above)
Comment (frederick-vs-ja) [reply tests/std/tests/P2443R1_views_chunk_by/test.cpp:243]: The name `Bool` looks terrible to me. Should it be `TrueLike`?

Comment (cpplearner) [reply tests/std/tests/P2443R1_views_chunk_by/test.cpp:243]: > The name `Bool` looks terrible to me. Should it be `TrueLike`?

Agreed that `Bool` is not a good name, but I think `TrueLike` would be misleading. The test only needs a *`boolean-testable`* type whose `operator!` returns a reference to a non-copy-constructible type. It doesn't matter whether its `operator bool` returns `true` or not.

Comment (hewillk) [reply tests/std/tests/P2443R1_views_chunk_by/test.cpp:243]: Is `NonCopyableBool?` better?

Comment (frederick-vs-ja) [reply tests/std/tests/P2443R1_views_chunk_by/test.cpp:243]: I want to use `PinnedSomething` for short. I think the following types have desired properties and model _`boolean-testable`_, but maybe too wordy.

```C++
class PinnedBool;

class PinnedFalse {
private:
    PinnedFalse() = default;
    PinnedFalse(const PinnedFalse&) = delete;
    PinnedFalse& operator=(const PinnedFalse&) = delete;

    friend class PinnedBool;

public:
    operator bool() {
        return false;
    }

    PinnedBool& operator!();
};

class PinnedBool : public PinnedFalse {
public:
    PinnedBool() = default;
    PinnedBool(const PinnedBool&) = delete;
    PinnedBool& operator=(const PinnedBool&) = delete;

    operator bool() {
        return true;
    }

    PinnedFalse& operator!() {
        return static_cast<PinnedFalse&>(*this);
    }
};

inline PinnedBool& PinnedFalse::operator!() {
    return static_cast<PinnedBool&>(*this);
}
```

Comment (CaseyCarter) [reply tests/std/tests/P2443R1_views_chunk_by/test.cpp:243]: `ImmoBool`! (As a rule, ignore naming suggestions from me.)

File: stl/inc/ranges [hunk 5]
```diff
@@ -6255,10 +6275,7 @@ namespace ranges {
             _STL_VERIFY(_Pred, "cannot increment a chunk_by_view iterator whose parent view has no predicate");
 #endif // _ITERATOR_DEBUG_LEVEL != 0
 
-            const auto _Not_pred = [&_Orig_pred = *_Pred]<class _Ty1, class _Ty2>(_Ty1&& _Left, _Ty2&& _Right) {
-                return !_STD invoke(_Orig_pred, _STD forward<_Ty1>(_Left), _STD forward<_Ty2>(_Right));
-            };
-            const auto _Before_next = _RANGES adjacent_find(_It, _RANGES end(_Range), _Not_pred);
+            const auto _Before_next = _RANGES adjacent_find(_It, _RANGES end(_Range), _Negated_pred<_Pr>{*_Pred});
```
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/ranges:6278]: Clang doesn't implement CTAD for aggregates.

**Review by CaseyCarter — APPROVED** [review]
Small suggestions, of which we have enough that I think I should push changes.

File: tests/std/tests/P2443R1_views_chunk_by/test.cpp (hunk 4 above)
Comment (CaseyCarter) [reply tests/std/tests/P2443R1_views_chunk_by/test.cpp]: I believe I've addressed the concerns in this discussion by moving `Bool` into the function that uses it and adding a comment that it should not be used in runtime code. Feel free to reopen discussion if you don't agree.

File: stl/inc/xutility [hunk 6]
```diff
@@ -640,30 +640,14 @@ template <class _Fn, class... _Its>
...
-
 template <indirectly_readable _It, indirectly_regular_unary_invocable<_It> _Proj>
 struct projected {
     using value_type = remove_cvref_t<indirect_result_t<_Proj&, _It>>;
-#if defined(__clang__) || defined(__EDG__)
-    indirect_result_t<_Proj&, _It> operator*() const;
-#else // ^^^ no workaround / workaround vvv
     indirect_result_t<_Proj&, _It> operator*() const {
-        _CSTD abort(); // TRANSITION, VSO-1308657
+        _CSTD abort();
```
Comment (CaseyCarter) [inline stl/inc/xutility:647]: For reviewers:

`projected` is a tool to simplify expressing the requirements of algorithms that use projections. The design intent is that the member functions don't have definitions - they are just signatures that exist for `iter_reference_t` et al. to look at. This hasn't worked well in practice in the real world, as witnessed by the warning suppressions for clang and MSVC around the definition, the additional bug report (VSO-1308657) mentioned there, and the fact that it broke yet again when I made `Bool` a local class in that PR. The PR uses `projected<Bool*, identity>`, and clang-cl decided to report an error about `operator*` being used-but-not-defined when it can't possibly be defined in another TU since `Bool` has internal linkage. Rather than trying to find yet another workaround for this perennial issue (four or five changes here in ~3 years?) I decided to give up and simply _always_ define `operator*` to call `abort`. The function should never be called in practice, so even if it does end up emitted into some object files the linker should be able to clean it up.

**Review by strega-nil-ms — APPROVED** [review]
I am sad about `projected::operator*`, but cools.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for fixing this compiler error and improving the ranges experience! :heart_eyes_cat: :lady_beetle: :tada:
